src/Plugin/Commerce/ShippingMethod/CanadaPost.php

```diff
@@ -171,10 +177,18 @@ public function submitConfigurationForm(array &$form, FormStateInterface $form_s
   public function calculateRates(ShipmentInterface $shipment) {
     $rates = [];
 
-    // Only attempt to collect rates if an address exits on the shipment.
+    // Only attempt to collect rates if an address exists on the shipment.
     if (!$shipment->getShippingProfile()->get('address')->isEmpty()) {
-      $this->canadapost_rate_service->setShipment($shipment);
-      $rates = $this->canadapost_rate_service->getRates();
+      $order_id = $shipment->order_id->target_id;
+      $order = Order::load($order_id);
+      /** @var \Drupal\commerce_store\Entity\Store $store */
+      $store = $order->getStore();
+      $address = $store->getAddress();
+      $originPostalCode = $address->getPostalCode();
+      $postalCode = $shipment->getShippingProfile()->address->postal_code;
+      $weight = $shipment->getWeight()->convert('g')->getNumber();
+      $response = $this->ratingService->getRates($originPostalCode, $postalCode, $weight);
+      $rates = $this->parseResponse($response);
     }
 
     return $rates;
```

Comment:
You should minimize code indentation where possible. This can be rewritten:

```
if ($shipment->getShippingProfile()->get('address')->isEmpty()) {
  return [];
}

...

return $rates;

Reply:
Also, the idea was to have a Drupal-specific service which would do all the calculations and return the prepared results. The caller (in this case the shipping method) would only have minimal code.

You should be passing the shipment to the a service function and that would everything, returing the prepared `ShippingRate[]` results.